File: Python/sudoku.py
import numpy as np
import math
import sys
grid1 = [[5,3,0,0,7,0,0,0,0],[6,0,0,1,9,5,0,0,0],[0,9,8,0,0,0,0,6,0],[8,0,0,0,6,0,0,0,3],[4,0,0,8,0,3,0,0,1],[7,0,0,0,2,0,0,0,6],[0,6,0,0,0,0,2,8,0],[0,0,0,4,1,9,0,0,5],[0,0,0,0,8,0,0,7,9]]
grid2 = [[".",".","4",".",".",".","6","3","."],[".",".",".",".",".",".",".",".","."],["5",".",".",".",".",".",".","9","."],[".",".",".","5","6",".",".",".","."],["4",".","3",".",".",".",".",".","1"],[".",".",".","7",".",".",".",".","."],[".",".",".","5",".",".",".",".","."],[".",".",".",".",".",".",".",".","."],[".",".",".",".",".",".",".",".","."]]
#print(np.matrix(grid2))
def isValidSudoku(board):
    """
        :type board: List[List[str]]
        :rtype: bool
    """
    #check cols
    for x in range(9):
        tempdict = {}
        for y in range(9):   
            #print(temparr)
            #print(board[y][x])
            #print(tempdict)
            if board[y][x] in tempdict:
                if board[y][x] != 0:
                    #print("colfail at " + str(x) + str(y))
                    return False
            tempdict[board[y][x]] = 1
    #check rows
    for y in range(9):
        tempdict = {}
        for x in range(9):  
            #print(board[y][x])
            #print(tempdict) 
            if board[y][x] in tempdict:
                if board[y][x] != 0:
                    #print("rowfail at " + str(y) + str(x))
                    return False
            tempdict[board[y][x]] = 1
    #check squares
    for alpha in range(3):
        for beta in range(3):    
            tempdict = {}
            for y in range(3*alpha, 3*alpha + 3):
                for x in range(3*beta, 3*beta + 3):
                    #print(board[y][x])
                    #print(tempdict) 
                    if board[y][x] in tempdict:
                        if board[y][x] != 0:
                            #print("squarefail at " + str(y) + str(x))
                            return False
                    tempdict[board[y][x]] = 1
    return True


def isPossible(y, x, n, grid):
    #global grid
    for i in range(0, 9):
        if grid[y][i] == n:
            return False
    for i in range(0, 9):
        if grid[i][x] == n:
            return False
    #check squares
    alpha = math.floor((y/3))*3
    beta = math.floor((x/3))*3
    for i in range(0,3):
        for j in range(0,3):
            if grid[alpha+i][beta+j] == n:
                return False
    #print("sup")
    return True

def checkChars(grid):
    dictionary  = {0: 0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0}
    for y in range(9):
        for x in range(9):
            if grid[y][x] not in dictionary:
                #print(grid[y][x])
                return -1
            else:
                dictionary[(grid[y][x])] += 1
    sum = 0
    for i in range(1,10):
        sum += dictionary[i]
    return sum

def outputSudoku(grid):
    with open("solutions.txt", 'a') as f:
        f.write("Solutions: \n")
        for i in range(9):
            for j in range(9):
                f.write(str(grid[i][j]) + " ")
            f.write(" \n")
    f.close()

def inputSudoku():
    #print("hi")
    grid = []
    with open("input.txt", 'r') as f:
        for line in f:
            grid.append(line)
    i = 0
    #print(grid)
    while i < len(grid):
        grid[i] = str(grid[i]).split()
        #print(grid[i])
        i += 1
    #print(grid)
    if len(grid) != 9:
        return False
    if len(grid[0]) != 9:
        return False
    for y in range(9):
        for x in range(9):
            if grid[y][x] == '.':
                grid[y][x] = 0
            else:
                grid[y][x] = int(grid[y][x])
    #print(grid)
    thingy = checkChars(grid)
    #print(thingy)
    if thingy < 0:
        print("Error in grid!\n")
        return False
    if thingy < 17:
        print("Multiple solutions may exist. Please check your grid.\n")
    truthiness = isValidSudoku(grid)
    if not truthiness:
        print("Invalid sudoku!\n")
        return False
    return grid


def solve(grid, counter):
    #if counter > 100:
    #    return
    for y in range(9):
        for x in range(9):
            if grid[y][x] == 0:
                for n in range(1,10):
                    if isPossible(y,x,n, grid):
                        grid[y][x] = n
                        if counter < 81:
                            solve(grid, counter + 1)
                        grid[y][x] = 0
                return
    print("hi")
    outputSudoku(grid)
    print(np.matrix(grid))
    #print(3/0)
    sys.exit()
    #input("hello")


def inToOut():
    grid = inputSudoku()
    print(grid)
    if grid != False:
        solve(grid, 0)

inToOut()

File: C/sudoku.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

bool isValidSudoku(int grid[][9]) {
    for(int x = 0; x < 9; x++) {
        int arr[9] = {0,0,0,0,0,0,0,0,0};
        int j = 0;
        for(int y = 0; y < 9; y++) {
            if (grid[y][x] == 0) {
                continue;
            }
            int i = 0;
            while (i < 9) {
                if (grid[y][x] == arr[i]) {
                    return false;
                }
                i++;
            }
            arr[j] = grid[y][x];
            j++;
        }
    }

    for(int y = 0; y < 9; y++) {
        int arr[9] = {0,0,0,0,0,0,0,0,0};
        int j = 0;
        for(int x = 0; x < 9; x++) {
            if (grid[y][x] == 0) {
                continue;
            }
            int i = 0;
            while (i < 9) {
                if (grid[y][x] == arr[i]) {
                    return false;
                }
                i++;
            }
            arr[j] = grid[y][x];
            j++;
        }
    }

    for(int alpha = 0; alpha < 3; alpha++) {
        for(int beta = 0; beta < 3; beta++) {
            int arr[9] = {0,0,0,0,0,0,0,0,0};
            int j = 0;
            int y = 3 * alpha;
            while (y < (3*alpha) + 3) {
                int x = 3*beta;
                while (x < (3*beta) + 3) {
                    if (grid[y][x] == 0) {
                        x++;
                        continue;
                    }
                    int i = 0;
                    while (i < 9) {
                        if (grid[y][x] == arr[i]) {
                            return false;
                        }
                        i++;
                    }
                    arr[j] = grid[y][x];
                    j++;

                    x++;
                }

                y++;
            }
        }
    }

    return true;
}

bool isPossible(int y, int x, int n,int grid[][9]) {
    for(int i = 0; i < 9; i++) {
        if (grid[y][i] == n) {
            return false;
        }
        if (grid[i][x] == n) {
            return false;
        }
    }
    int alpha = (y/3) *3;
    int beta = (x/3) *3;
    for(int i = 0; i < 2; i++) {
        for(int j = 0; j < 2; j++) {
            if (grid[alpha+i][beta+j] == n){
                return false;
            }
        }
    }

    return true;
}

int checkChars (int **grid) {
    int dict[10] = {0,0,0,0,0,0,0,0,0,0};
    for (int y = 0; y < 9; y++) {
        for (int x = 0; x < 9; x++) {
            //whatever gonna hardcode this because anything other than 0 to 9 we return -1 anyway
            if (grid[y][x] == 0) {
                dict[0]++;
            } else if (grid[y][x] == 1) {
                dict[1]++;
            } else if (grid[y][x] == 2) {
                dict[2]++;
            } else if (grid[y][x] == 3) {
                dict[3]++;
            } else if (grid[y][x] == 4) {
                dict[4]++;
            } else if (grid[y][x] == 5) {
                dict[5]++;
            } else if (grid[y][x] == 6) {
                dict[6]++;
            } else if (grid[y][x] == 7) {
                dict[7]++;
            } else if (grid[y][x] == 8) {
                dict[8]++;
            } else if (grid[y][x] == 9) {
                dict[9]++;
            } else {
                return -1;
            }
        }
    }
    int sum = 0;
    for(int i = 0; i < 10; i++) {
        sum += dict[i];
    } 
    return sum;
}

int asciiToDig(int character) {
    /* hard code in digs from 0 to 9, just return 0 if its not a digit in that range*/
    if (character <= 48) {
        return 0;
    } else if (58 <= character) {
        return 0;
    } else {
        return (character - 48);
    }
}

void inputSudoku(int grid[][9]) {
    FILE *fptr = fopen("duplicate.txt", "r");
    if (fptr == NULL) {
        printf("Invalid Input File!");
        exit(1);
    }
    for(int y = 0; y < 9; y++) {
        int i = 0;
        for(int x = 0; x < 18; x++) {
            char c[2];
            fscanf(fptr, "%c", c);
            if (x % 2 == 0){
                int d = *c;
                //printf("%d ", d);
                int e = asciiToDig(d);
                //printf("%d ", e);
                grid[y][i] = e;
                //printf("%d ", grid[y][i]);
                i++;
            }
        }
        
        /*printf("\n");
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++){
                printf("%d ", grid[i][j]);
        }
        printf("\n");
    }*/

    }
    fclose(fptr);
    
}

void outputSudoku(int grid[][9]) {
    FILE *fptr = fopen("output.txt", "a");
    if (fptr == NULL) {
        printf("Invalid Output File!");
        exit(1);
    }
    //printf("hello");
    fprintf(fptr, "Solution: \n");
    for (int x = 0; x < 9; x++) {
        for (int y = 0; y < 9; y++) {
            //printf("hello \n");
            fprintf(fptr, "%d ", grid[x][y]);
        }
        fprintf(fptr, " \n");
    }
    fclose(fptr);
}

void solve(int grid[][9]) {
    for(int y= 0 ; y < 9; y++) {
        for(int x =0; x < 9; x++){
            //printf("hasn't crashed here\n");
            if (grid[y][x] == 0) {
                for(int n = 1; n < 10; n++) {
                    if (isPossible(y,x,n,grid)) {
                        //printf("%d %d \n", y, x);
                        grid[y][x] = n;
                        //outputSudoku(grid);
                        solve(grid);
                        grid[y][x] = 0;
                    }
                }
                return;
            }
        }
    }
    printf("sup\n");
    outputSudoku(grid);


    //exit(0); //only one solution. Too bad!
}


int main() {
    FILE *fptr = fopen("output.txt", "a");
    fprintf(fptr, "Solution: \n");
    fclose(fptr);
    int potato[][9] = {{5,3,0,0,7,0,0,0,0},{6,0,0,1,9,5,0,0,0},{0,9,8,0,0,0,0,6,0},{8,0,0,0,6,0,0,0,3},{4,0,0,8,0,3,0,0,1},{7,0,0,0,2,0,0,0,6},{0,6,0,0,0,0,2,8,0},{0,0,0,4,1,9,0,0,5},{0,0,0,0,8,0,0,7,9}};
    int (*ptr)[]  = potato;
    //printf("hi \n");
    //inputSudoku(ptr);
    /*bool poss = isPossible(4,4,4,ptr);
    printf("%d ", poss);*/
    solve(ptr);
    /*
    for (int y = 0; y < 9; y++) {
        for(int x = 0; x < 9; x++) {
            printf("%d ", potato[y][x]);
        }
        printf("\n");
    }*/
    //outputSudoku(ptr);
}